### PR TITLE
Support not in package dependencies constraints

### LINKED ISSUE #11205: package constraints in dune-project does not support not (!)
## Expected Behavior

I want to express the opam constraint: "foo" {with-test & >= "2.0" | ! with-test & >= "1.0"}

I can't since "not" nor "!" are supported.

## Actual Behavior

Error: Unknown constructor !

## Reproduction

in your dune-project
```
(package
 (depends
  (ocp-indent
   (or
    (and
     (! :with-test)
     (>= 1.0))
    (and
     :with-test
     (>= 2.0))))
```

### PR DESCRIPTION
**Pull request by art-w**
Resolves #11205

### CODE REVIEW

File: test/blackbox-tests/test-cases/dune-project-meta/github11205.t [hunk 1]
```diff
@@ -0,0 +1,64 @@
+Support for negation in package dependencies:
+
+  $ cat > dune-project <<EOF
+  > (lang dune 3.18)
+  > (generate_opam_files)
+  > (package
+  >   (name foo)
+  >   (allow_empty)
+  >   (depends
+  >    (ocp-indent
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/dune-project-meta/github11205.t:10]: Can you add a test when we have a package named "not" with a constraint? E.g. `(not (>= 3.0))`

Comment (art-w) [reply test/blackbox-tests/test-cases/dune-project-meta/github11205.t:10]: Done!

File: test/blackbox-tests/test-cases/dune-project-meta/github11205.t [hunk 2]
```diff
@@ -0,0 +1,66 @@
+Support for negation in package dependencies:
```
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/dune-project-meta/github11205.t:1]: Given this is a new feature, maybe this test could be added in some other existing cram test that is testing the other operators?

My concern is that `github11205` is not very descriptive and if we need to add a test on some specific behavior we might forget that this already exists and then duplicate it.

File: test/blackbox-tests/test-cases/dune-project-meta/github11205.t [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  >    (ocp-indent
+  >     (not :with-test))))
+  > EOF
+  $ dune build
+  File "dune-project", line 8, characters 4-20:
+  8 |     (not :with-test))))
+          ^^^^^^^^^^^^^^^^
+  Error: Not operator is only available since version 3.18 of the dune
+  language. Please update your dune-project file to have (lang dune 3.18).
+  [1]
```
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/dune-project-meta/github11205.t:66]: I think it would be also good to show that `dune pkg lock` correctly handles `not` and will avoid adding a package that has a `not` constraint (and error out if there is no solution).

Probably not in this cram test but in the dune pkg cram tests.

File: src/dune_lang/package_constraint.ml [hunk 4]
```diff
@@ -138,6 +144,13 @@ let decode =
       ; ( "or"
         , let+ x = logical_op t in
           Or x )
+      ; ( "not"
+        , let+ x = t
+          and+ loc = loc
```
Comment (Leonidas-from-XIV) [inline src/dune_lang/package_constraint.ml:149]: Small nitpick but if `version` is `>= (3, 18)` then we don't need `loc`.

File: doc/changes/11404.md [hunk 5]
```diff
@@ -0,0 +1 @@
+- Support not in package dependencies constraints (#11404, @art-w, reported by @hannesm)
```
Comment (Leonidas-from-XIV) [inline doc/changes/11404.md:1]:
```suggestion
- Support `not` in package dependencies constraints (#11404, @art-w, reported by @hannesm)
```

**Review by Leonidas-from-XIV — COMMENTED** [review]
Nice and the code looks pretty good. My concerns are mostly about the tests which could be a bit more discoverable.

**Review by rgrinberg — APPROVED** [review]

**Comment by Leonidas-from-XIV** [comment]
@art-w Ping?

File: test/blackbox-tests/test-cases/dune-project-meta/github11205.t (hunk 3 above)
Comment (art-w) [reply test/blackbox-tests/test-cases/dune-project-meta/github11205.t]: Good point, I added some negation tests for `dune pkg lock` :)

File: test/blackbox-tests/test-cases/dune-project-meta/github11205.t (hunk 2 above)
Comment (art-w) [reply test/blackbox-tests/test-cases/dune-project-meta/github11205.t]: I moved the negation tests to the existing `binops.t` and renamed it `operators.t`

**Review by Leonidas-from-XIV — APPROVED** [review]
